jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletContextHandler.java

```diff
@@ -1493,4 +1501,43 @@ public void destroy(Object o)
             decorator.destroy(o);
         }
     }
+
+    /**
+     * A utility class to hold a {@link ServletContainerInitializer} and implement the
+     * {@link ServletContainerInitializerCaller} interface so that the SCI is correctly
+     * started if an instance of this class is added as a bean to a {@link ServletContextHandler}.
+     */
+    public static class Initializer extends AbstractLifeCycle implements ServletContainerInitializerCaller
+    {
+        private final ServletContextHandler _context;
+        private final ServletContainerInitializer _sci;
+        private final Set<Class<?>> _classes;
+
+        public Initializer(ServletContextHandler context, ServletContainerInitializer sci, Set<Class<?>> classes)
+        {
+            _context = context;
+            _sci = sci;
+            _classes = classes;
+        }
+
+        public Initializer(ServletContextHandler context, ServletContainerInitializer sci)
+        {
+            this(context, sci, Collections.emptySet());
+        }
+
+        @Override
+        protected void doStart() throws Exception
+        {
+            boolean oldExtended = _context.getServletContext().isExtendedListenerTypes();
+            try
+            {
+                _context.getServletContext().setExtendedListenerTypes(true);
+                _sci.onStartup(_classes, _context.getServletContext());
+            }
+            finally
+            {
+                _context.getServletContext().setExtendedListenerTypes(oldExtended);
+            }
+        }
+    }
```

Comment:
This utility class avoids the need to create custom SCI callers.   The name is not perfect, but this name space is a bit messed up.  It does read well when used.

Reply:
If an SCI is being run on a ServletContextHandler, there's no expectations of bytecode scanning (IMO).

It would be sweet if we could just do ...

``` java
ServletContextHandler handler = new ServletContextHandler();
handler.addBean(new MyLibServletContextInitializer());
handler.addBean(new org.jboss.weld.environment.servlet.EnhancedListener()); // this is an SCI
server.setHandler(handler);
server.start(); // this loads the SCI's from the beans and uses them.
```

Just have ServletContextHandler be aware of SCI as beans and load / call them with no/empty classes.

Or, if this raw/naked SCI as bean concept isn't right, how about a specific call on ServletContextHandler for them?

``` java
ServletContextHandler handler = new ServletContextHandler();
handler.addInitializer(new MyLibServletContextInitializer(), Collections.emptySet());
handler.addInitializer(new org.jboss.weld.environment.servlet.EnhancedListener(), Collections.emptySet());
server.setHandler(handler);
server.start();
```

Reply:
@joakime if we were adding this mechanism from scratch, I agree that just adding the SCIs as beans would be simplest.
However, we already have the concept of a ServletContainerInitializerCaller set as a bean and it is used at least a little bit.... so adding a utility instance of that feels simpler than having two mechanisms.     I guess we could make `Initializer` private and check for SCI beans... so it would be a feeder to the existing mechanism.... torn.... @janbartel thoughts?

Reply:
You can't do away with some kind of wrapper for the SCIs because the SCI.onStartup() method can have the list of classes passed in, so you need to remember somewhere what they are.

Reply:
Ah yes - naked SCIs only work if they have no classes to be passed in.  Whilst that will be rare for the usage intended, it still would mean another mechanism, so let's just keep it simple.